### PR TITLE
Merge go-crypto into tendermint

### PR DESCRIPTION
**Pull request by Liamsi**
* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
~~* [ ] Wrote tests~~ doesn't apply
~~* [ ] Updated CHANGELOG.md~~

Part of #1309
Contains ~~a lot~~ basically all of changes from #1721

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -74,7 +74,7 @@ Additional information about some - and eventually all - of the sub-projects bel
 
 * [ABCI](http://github.com/tendermint/abci), the Application Blockchain Interface
 * [Go-Wire](http://github.com/tendermint/go-wire), a deterministic serialization library
-* [Go-Crypto](http://github.com/tendermint/go-crypto), an elliptic curve cryptography library
+* [Go-Crypto](http://github.com/tendermint/tendermint/crypto), an elliptic curve cryptography library
```
Comment (ValarDragon) [inline README.md]: Is "an elliptic curve cryptography library" the correct epithet for this?

File: rpc/lib/server/handlers_test.go [hunk 2]
```diff
@@ -19,13 +18,6 @@ import (
 	"github.com/tendermint/tmlibs/log"
 )
 
-//////////////////////////////////////////////////////////////////////////////
```
Comment (melekes) [inline rpc/lib/server/handlers_test.go]: please return back

File: rpc/lib/server/handlers_test.go [hunk 3]
```diff
@@ -116,44 +108,3 @@ func TestUnknownRPCPath(t *testing.T) {
 	// Always expecting back a 404 error
 	require.Equal(t, http.StatusNotFound, res.StatusCode, "should always return 404")
 }
-
-//////////////////////////////////////////////////////////////////////////////
```
Comment (melekes) [inline rpc/lib/server/handlers_test.go]: please return back

File: types/canonical_json.go [hunk 4]
```diff
@@ -3,14 +3,13 @@ package types
...
 
-	"github.com/tendermint/go-amino"
 	cmn "github.com/tendermint/tmlibs/common"
 )
 
 // Canonical json is amino's json for structs with fields in alphabetical order
 
 // TimeFormat is used for generating the sigs
-const TimeFormat = amino.RFC3339Millis
+const TimeFormat = "2006-01-02T15:04:05.000Z"
```
Comment (melekes) [inline types/canonical_json.go]: old version `const TimeFormat = amino.RFC3339Millis` is correct

File: types/priv_validator.go [hunk 5]
```diff
@@ -63,22 +63,31 @@ func (pv *MockPV) GetPubKey() crypto.PubKey {
...
 	proposal.Signature = sig
 	return nil
 }
 
 // signHeartbeat signs the heartbeat without any checking.
 func (pv *MockPV) SignHeartbeat(chainID string, heartbeat *Heartbeat) error {
-	sig := pv.privKey.Sign(heartbeat.SignBytes(chainID))
+	sig, err := pv.privKey.Sign(heartbeat.SignBytes(chainID))
+	if err != nil {
+		panic(err)
```
Comment (melekes) [inline types/priv_validator.go]: `return err`

File: types/priv_validator.go [hunk 6]
```diff
@@ -63,22 +63,31 @@ func (pv *MockPV) GetPubKey() crypto.PubKey {
...
 	return nil
 }
 
 // Implements PrivValidator.
 func (pv *MockPV) SignProposal(chainID string, proposal *Proposal) error {
 	signBytes := proposal.SignBytes(chainID)
-	sig := pv.privKey.Sign(signBytes)
+	sig, err := pv.privKey.Sign(signBytes)
+	if err != nil {
+		panic(err)
```
Comment (melekes) [inline types/priv_validator.go]: `return err`

File: p2p/conn/secret_connection.go [hunk 7]
```diff
@@ -267,7 +267,10 @@ func genChallenge(loPubKey, hiPubKey *[32]byte) (challenge *[32]byte) {
 }
 
 func signChallenge(challenge *[32]byte, locPrivKey crypto.PrivKey) (signature crypto.Signature) {
-	signature = locPrivKey.Sign(challenge[:])
+	signature, err := locPrivKey.Sign(challenge[:])
+	if err != nil {
+		panic(err)
```
Comment (melekes) [inline p2p/conn/secret_connection.go]: can you add a todo to modify func to return err?

File: types/canonical_json.go (hunk 4 above)
Comment (Liamsi) [reply types/canonical_json.go]: The version of amino this uses doesn't contain the constant anymore:
https://github.com/tendermint/go-amino/commit/f81fa38144591e3e0581724fc30e5db820aba1a5

File: p2p/conn/secret_connection.go (hunk 7 above)
Comment (Liamsi) [reply p2p/conn/secret_connection.go]: Sure! Will do (and also open an issue so that doesn't get lost)

**Comment by Liamsi** [comment]
Thanks for the review!! Addressed all your comments @melekes.

File: types/canonical_json.go (hunk 4 above)
Comment (melekes) [reply types/canonical_json.go]: oh, sorry. this one --->>>
https://github.com/tendermint/tendermint/pull/1756/files#diff-2a5c02592829cd2adae17d586208b772

File: rpc/lib/server/handlers_test.go [hunk 8]
```diff
@@ -108,3 +116,44 @@ func TestUnknownRPCPath(t *testing.T) {
...
+
+//////////////////////////////////////////////////////////////////////////////
+// JSON-RPC over WEBSOCKETS
+
+func TestWebsocketManagerHandler(t *testing.T) {
+	s := newWSServer()
+	defer s.Close()
+
+		// check upgrader works
+			d := websocket.Dialer{}
```
Comment (melekes) [inline rpc/lib/server/handlers_test.go]: what's up with formatting here?

File: types/canonical_json.go (hunk 4 above)
Comment (Liamsi) [reply types/canonical_json.go]: Tests fail, if I do this change.

File: rpc/lib/server/handlers_test.go (hunk 8 above)
Comment (Liamsi) [reply rpc/lib/server/handlers_test.go]: Somehow I disabled gofmt. Thanks

**Comment by Liamsi** [comment]
> Reviewed against #1721, LGTM.

Thank you too @cwgoes 👍

File: config/config.go [hunk 9]
```diff
@@ -279,8 +279,9 @@ type P2PConfig struct {
 	// Time to wait before flushing messages out on the connection, in ms
 	FlushThrottleTimeout int `mapstructure:"flush_throttle_timeout"`
 
-	// Maximum size of a message packet payload, in bytes
-	MaxPacketMsgPayloadSize int `mapstructure:"max_packet_msg_payload_size"`
+	// Maximum size of a message packet, in bytes
+	// Includes a header, which is ~13 bytes
+	MaxPacketMsgSize int `mapstructure:"max_packet_msg_size"`
```
Comment (ebuchman) [inline config/config.go]: breaking

File: types/heartbeat_test.go [hunk 10]
```diff
@@ -37,11 +39,11 @@ func TestHeartbeatWriteSignBytes(t *testing.T) {
 	hb := &Heartbeat{ValidatorIndex: 1, Height: 10, Round: 1}
 	bz := hb.SignBytes("0xdeadbeef")
 	// XXX HMMMMMMM
-	require.Equal(t, string(bz), `{"@chain_id":"0xdeadbeef","@type":"heartbeat","height":10,"round":1,"sequence":0,"validator_address":"","validator_index":1}`)
+	require.Equal(t, string(bz), `{"@chain_id":"0xdeadbeef","@type":"heartbeat","height":"10","round":"1","sequence":"0","validator_address":"","validator_index":"1"}`)
 
 	plainHb := &Heartbeat{}
 	bz = plainHb.SignBytes("0xdeadbeef")
-	require.Equal(t, string(bz), `{"@chain_id":"0xdeadbeef","@type":"heartbeat","height":0,"round":0,"sequence":0,"validator_address":"","validator_index":0}`)
+	require.Equal(t, string(bz), `{"@chain_id":"0xdeadbeef","@type":"heartbeat","height":"0","round":"0","sequence":"0","validator_address":"","validator_index":"0"}`)
```
Comment (ebuchman) [inline types/heartbeat_test.go]: so ints are strings now huh

**Pull request merged**